src/app/_components/Hint.js

```diff
@@ -0,0 +1,90 @@
+'use client';
+
+import { cloneElement, createContext, useContext, useEffect, useState } from 'react';
+import PropTypes from 'prop-types';
+import { cn } from '@/utils/cn';
+import { ClientPortal } from './ClientPortal';
+
+const HintContext = createContext();
+
+export function Hint({ children }) {
+  const [openName, setOpenName] = useState('');
+  const close = () => setOpenName('');
+  const open = setOpenName;
+
+  return <HintContext.Provider value={{ open, close, openName }}>{children}</HintContext.Provider>;
+}
+
+export function Show({ children, opens: opensWindowName, actions }) {
+  const { open, close } = useContext(HintContext);
+
+  // close hints when window is scrolled
+  useEffect(() => {
+    const handleScroll = () => {
+      close();
+    };
+
+    window.addEventListener('scroll', handleScroll);
+
+    return () => window.removeEventListener('scroll', handleScroll);
+  }, [close]);
+
+  return cloneElement(children, {
+    id: opensWindowName,
+    onClick: e => {
+      e.stopPropagation();
+      actions?.onClick?.({ open, close });
+    },
+    onMouseEnter: e => {
+      e.stopPropagation();
+      actions?.onMouseEnter?.({ open, close });
+    },
+    onMouseLeave: e => {
+      e.stopPropagation();
+      actions?.onMouseLeave?.({ open, close });
```

Comment:
You can add logic here instead of delegating it to the parent component, so instead of actions?.onClick?.({ open, close }); use the toggle method(export it from the context along with open and close methods), instead of  actions?.onMouseEnter?.({ open, close }); use the method open, and replace actions?.onMouseLeave?.({ open, close }); with the method close

Reply:
Thanks for comment! If i understood you correctly, i need to leave actions just in ShowHint component. At the moment, we need extra login to define when hint should be show, so i defined in just in ShowHint.